README
NAME
    Test::System - Test suite oriented for testing system administration
    tasks

SYNOPSIS
        use Test::System;

        my $suite = Test::System->new(
                format => 'consoletable',
                nodes => 'example.com',
                test_groups => '~/code/my/system/tests.yaml'
                );
        $suite->runtests;

DESCRIPTION
    Loads and runs a set of tests for verifying a system administration task
    before a possible incident happens or after a task is executed.

    Tests can be run as perl scripts and actually in any format TAP::Harness
    supports.

    Some examples of how Test::System can help you:

    *   Before a deployment probably you want to make sure your systems and
        configuration files are ready to go. Also you probably want to make
        sure that if you needed to apply patches to your configuration
        files, they really got applied (even in a development server or
        verify directly on production servers). So here Test::System can
        help you verifying this in the form of a test case. Of course, you
        write the test case :-)

    *   What about if you have many servers and although you have nagios to
        verify the speed of the CPU sometimes you need to verify it
        "on-demand" (perhaps when you get new servers in your datacenter).
        Here Test::System can help you to verify that these servers are
        ready to go before even they get connected to nagios or your
        favorite monitoring tool.

    One of the things you need to keep in mind is that Test::System is NOT a
    monitoring system. It will let you verify things on demand, things that
    are repetitive and perhaps you want them to be automated.

Attributes
    The module offers a list of attributes, some of them are read-only.

    test_groups
          YAML filename that has a list of available tests. This is not
          required but can be useful if you want to group tests (like
          *hardware.yaml*, *net.yaml*, etc). Comes also handy when user does
          not provide a list of tests to execute, so all the tests listed in
          this file are executed.

          An example of a YAML file:

              ping:
                  description: Test the ping and do foo and make bar
                  code: test/foo.pl
              cpu:
                  description: Test the CPU of nodes
                  code: test/cpu.pl

    available_tests
          Is a read only attribute that contains a list of all available
          tests found in the YAML file provided by "test_groups".

    nodes Is an attribute that can be represented as a string (like a
          hostname) or as a list (where each item will be a node/hostname).
          This attribute has write access and is where the tests are going
          to be executed to.

          Another way of setting this value is when "runtests()" is called,
          a test plan (in the form of a YAML file) can be provided and it
          can contain the list of nodes to use.

    format
          A write access string that has the format of how the tests should
          be presented, please refer to the modules available under
          Test::System::Output or in your custom factory
          ("format_factory_class" attribute) class if this is the case.

    available_formats
          A list of available formats, read only.

    format_factory_class
          If you want to use your own Factory for creating your output you
          can set this to your class name (NOT the object).

          Please note that Test::System will continue to use
          Test::System::Output::Factory, the reason of this is that any
          factory subclasses should add/register their classes via their
          parent class. So in this case all your new format classes should
          be added to Test::System::Output::Factory. For more information
          please take a look to Class::Factory.

    harness
          Is (or will be) the harness instance once "runtests" is executed.

    parameters
          An attribute with write access permission. This attribute will
          transform all the items of this hash to environment variables.

          The use of this attribute is very handy if you want to provide
          some additional data for your tests and since the tests are run in
          separate forks with Test::Harness then the only possible way to
          keep them is to make them available through the environment (%ENV
          hash).

          Please be warned that only scalars are stored in environment
          variables, those that are an array will be converted to CSV values
          while the rest of the data types will be lost.

          In your tests if you want to use any of these parameters they will
          be available through the environment variables with a prefix of:
          "TEST_SYSTEM_" or you can use Test::System::Helper to get their
          values.

    results
          A hash reference that contains the results of the tests. This
          information is generated by the TAP::Parser::Aggregator.

          * passed: Total of passed tests
          * total: Total number of tests (passed or not)
          * skipped: Total of skipped tests
          * failed: Total of failed tests.

    status
          A string (read-only) that contains a word that describes the
          status of all the tests. This value is also generated by
          TAP::Parser::Aggregator.

    show_warnings
          By default we show warnings of all things that can make your tests
          to run in a different way than expected.

Methods
  runtests(@tests,%options)
    It will run the given test cases (@tests). The @tests can be:

    * An array of test files to execute.
    * Or, a string pointing to a test plan file (YAML).

    If no @tests are given but we have a "test_groups" then ALL the tests
    listed inside this file will be executed.

    The %options is a hash of options that will be passed to the
    TAP::Harness object, some useful parameters are:

    * verbosity
        By default we mute everything with -9.

    * color
        If you want the output (in console) to have color

    * formatter
        Although we use Test::System::Output::Factory to offer a set of
        formatters you can provide your own formatter object. See
        "available_formats".

    * jobs
        If you have many tests you probably want to increment this value
        (that defaults to 1) so other tests can be run at the same time.

  run_test_plan($test_plan, %options)
    It loads the given test plan file (should be written in YAML) in order
    to get a list of tests to execute. Once it has a list of tests then it
    calls "runtests" with this list and with the %options.

  setup_parser($parser)
    This method should never be called directly since this is triggered by
    TAP::Harness when every TAP::Parser object gets created.

    This method is useful for setting the callbacks we want the parser to
    trigger.

    $parser should be a TAP::Parser object.

  pretests_verification(@tests, %options)
    Does some verification before the tests are executed. This method gets
    called by "runtests".

    The parameters it accepts are the same parameters passed to "runtests"
    with the main difference the tests are already filtered (eg, if a test
    plan in YAML was provided we will use it or we will use all the tests
    listed inside the "test_groups".

  what_happened($parser, $result_test)
    Similar to *setup_parser*, this should never be called since it will be
    triggered by TAP::Parser after each (sub)tests gets executed.

    By default we check the reason of why the test failed so later we
    provide a simple hash to parse with all the tests and the reasons of why
    they failed (if this could be the case). And because we need to know
    what TAP::Parser instance we are processing we need to ask for it as a
    first parameter.

    Parameters:

    * $parser should be a TAP::Parser
    * $result_test should be a TAP::Parser::Result::Test.

  set_results($aggregator)
    This is a callback and is only called/triggered by TAP::Harness when all
    the tests are done. It will fill the "results" attribute with
    information of everything.

    Refer to the documentation of *results* for more information.

    The paramenter $aggregator is a TAP::Parser::Aggregator.

  prepare_environment()
    Prepares the environment by settings the needed environment values so
    they can be used later by the tests

  clean_environment()
    Cleans/deletes all the environment variables that match "TEST_SYSTEM_*"

  get_tests_from_test_plan($yaml_file,
        $do_not_fill_parameters, 
        $do_not_fill_nodes)
    Reads the given tests yaml file ($yaml_file). This YAML file should have
    at least a list of tests (in the form of a hash) and optionally can also
    have parameters the tests should contain.

    Although this method is used mostly internally there's the option to
    call it as any other method *Test::System* offers.

    By default it will fill the parameters of your *Test::System* instance
    but by passing $do_not_fill_parameters (second parameter) as true or
    something that Perl understands as true then it will skip the part. This
    should be presented as a hash in YAML syntax.

    The above apply also to $do_not_fill_nodes (third parameter). This
    should be presented as an array in YAML syntax.

    Once the file is read an array with all the tests will be returned.

    An example of a YAML test plan file can be found inside the "examples"
    directory or:

        tests:
            - ping
            - cpu
            - memory
        parameters:
            foo: bar
            bar: zoo
        nodes:
            - pablo.com.mx
            - example.com

SEE ALSO
    Take a look to an awesome and pretty similar CPAN module: Test::Server.

AUTHOR
    Pablo Fischer (pablo@pablo.com.mx).

COPYRIGHT
    Copyright (C) 2009 by Pablo Fischer.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



CHANGES
Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 22:23:06 2009 -0600

    Now each test can be represented as a scalar (the name/filepath) or as
    an array (of 2 items, first item of filepath and second for the
    description). The same applies to test plans.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 19:47:23 2009 -0600

    To the details of each test we now have a new key: is_ok to know if the
    test failed or passed.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 15:49:15 2009 -0600

    New changes

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 15:47:59 2009 -0600

    When a new format factory gets added we should still be calling
    Test::system::Output::Factory cause this is where _all_ formats are.
    Then user can override a format type by using Class::Factory.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 13:19:48 2009 -0600

    New changes

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 23 13:18:05 2009 -0600

    Cleaner way: runtests for an array of test and run_test_plan for
    running/loading a test plan (in yaml), then it will call runtests() with
    the array of test to execute.
    
    Also now Test::System makes sure that %options is really a hash, if it
    doesn't then it dies.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Sun Nov 22 13:51:09 2009 -0600

    Updated CPAN files and examples too

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Sat Nov 21 19:16:42 2009 -0600

    Files needed for building a new version

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Sat Nov 21 19:15:20 2009 -0600

    New changes.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Sat Nov 21 19:11:26 2009 -0600

     - Catch notes that come from the tests and place them on
     $result->{'notes'}
     - Result details are moved form $result->{'result'} to
     $result->{'details'}
     - The custom_factory attribute is now called format_factory_class
     - Fix some bugs when we send a plan rather than an array of @tests.
     - By default warnings are printed, they can be disabled by setting
     Test::System->show_warnings to something Perl understand as false
     (empty, 0, undef)
     - For examples/suites.pl now it is a CLI command that uses
     Getopt::Awesome (another of my *awesome* modules). This is not a
     required dependency but instead it is a "recommended" dependency.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 20 11:36:44 2009 -0800

    Extra CPAN file updates

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 20 11:35:09 2009 -0800

    Some other undef fixes

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 20 00:57:16 2009 -0800

    CPAN files

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 20 00:55:48 2009 -0800

    Fixed reference bugs

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Thu Nov 19 19:04:13 2009 -0800

    New version of files before I create the tarball

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Thu Nov 19 19:00:55 2009 -0800

    A fiix, we shouldn't be depending on Universal::isa

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Wed Nov 18 11:13:46 2009 -0800

    Documentation improvements and add a new function: pretest_verification

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 22:21:43 2009 -0800

    Update METa.yml

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 22:20:00 2009 -0800

    Update CHANGES, Makefile and README with latest stuff

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 22:18:03 2009 -0800

    Now Test::System can connect with callbacks of TAP::Parser and
    TAP::Harness in order to know whats really happening and to build a hash
    with all this information so user can later process.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 12:54:24 2009 -0800

    Update the CHANGES file and the README

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 12:53:29 2009 -0800

    Increment version (forgot a file)

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 12:13:30 2009 -0800

    Fixed some documentation issues and improved it.

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 02:21:48 2009 -0800

    Increment version since I did something stupid via PAUSE

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 02:19:12 2009 -0800


Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 02:12:45 2009 -0800

    ./Makefile.PL result

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Fri Nov 13 02:07:02 2009 -0800

    Let Test::System read a test plan from a YAML file and execute from
    there.
    
    Also drop some attributes from Test::System since they are not really
    needed or some are not that *required*

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 9 14:05:18 2009 -0800

    Make it more CPAN standard

Author: Pablo Fischer <pablo@pablo.com.mx>
Date:   Mon Nov 9 09:53:02 2009 -0800

    Initial revision


Test/System.pm
#
# Test::System
#
# Author(s): Pablo Fischer (pfischer@cpan.org)
# Created: 11/07/2009 17:36:17 PST 17:36:17
package Test::System;

=head1 NAME

Test::System - Test suite oriented for testing a *system*

=head1 SYNOPSIS

    use Test::System;

    my $suite = Test::System->new(
            format => 'consoletable',
            nodes => 'example.com',
            tests_yaml => '~/code/my/system/tests.yaml'
            );
    $hung->runtests;

=head1 DESCRIPTION

Loads and runs the available tests cases located in the namespace of
Seco::Insanity::Tests. User can specify if only one or a group of tests cases
should be run.

=cut

use Moose; # it turns strict and warnings
use File::Basename qw(dirname);
use YAML::Syck;
use Test::System::Output::Factory;
use TAP::Harness;
use Data::Dumper;

=head1 Attributes

Test::System exports a number of attributes that some only have read only
access and others allow write access.

=over 6

=item B<tests_yaml>

YAML filename of where a list of available tests are. You probably want to take
a look to C<examples/tests.yaml> but a quick example is:

    ping:
        description: Test the ping and do foo and make bar
        code: test/foo.pl
    cpu:
        description: Test the CPU of nodes
        code: test/cpu.pl

=cut
has 'tests_yaml' => (
        is => 'rw',
        isa => 'Str',
        required => 1,
        trigger => \&_read_tests_yaml_file
        );

=item B<tests_dir>

Read only attribute generated by the read of the YAML file. Its value is the
directory path of where C<tests_yaml> is located.

=cut
has 'tests_dir' => (
        is => 'ro',
        isa => 'Str'
        );

=item B<available_tests>

Is a read only attribute that contains a list of all available tests found in
the YAML file provided by tests_yaml.

=cut
has 'available_tests' => (
        is => 'ro',
        isa => => 'ArrayRef[Str]',
        );

=item B<nodes>

Is an attribute that can be represented as a string (like a hostname) or as a
list (where each item will be a node/hostname). This attribute has write access
and is where the tests are going to be executed to.

=cut
has 'nodes' => (
        is => 'rw',
        isa => 'Any',
        required => 1,
        trigger => \&_verify_nodes_datatype
        );

=item B<format>

A write access string that has the format of how the tests should be presented,
please refer to the modules available under Test::System::Output

=cut
has 'format' => (
        is => 'rw',
        isa => 'Str',
        default => 'console',
        trigger => \&_verify_format
        );

=item B<available_formats>

A list of available formats, read only.

=cut
has 'available_formats' => (
        is => 'ro',
        isa => => 'HashRef',
        default => \&_generate_output_types_hash
        );

=item B<parameters>

An attribute with write access permission. This attribute will transform all
the items of this hash to environment variables.

The use of this attribute is very handy if you want to provide some additional
data for your tests and since the tests are run in separate forks with
Test::Harness then the only possible way to keep them is to make them available
through the environment (C<%ENV> hash).

Please be warned that only scalars are stored in environment variables, those
that are an array will be converted to CSV values while the rest of the data
types will be lost.

In your tests if you want to use any of these parameters they will be available
through the environment variables with a prefix of: I<TEST_SYSTEM_>.

=cut
has 'parameters' => (
        is => 'rw',
        isa => 'HashRef[Str]',
        );

=back

=head1 Methods

=over 4

=item B<runtests( @tests , %options)>

It will run a group of given test cases, however if no list is given or is
empty then all the available cases will be run.

The C<%options> is a hash of options that will be passed to the B<TAP::Harness>
object, some useful parameters are:

=over 4

=item * verbosity

By default we mute everything with C<-9>.

=item * color

If you want the output (in console) to have color

=item * formatter

Although we use B<Test::System::Output::Factory> to offer a set of formatters
you can provide your own formatter object.

=item * jobs

If you have many tests you probably want to increment this value (that defaults
to C<1>) so other tests can be run at the same time.

=back

=cut
sub runtests {
    my ($self, $tests, $options) = @_;
    my @tests_to_run = keys(%{$self->available_tests}) unless
        scalar $tests eq 0;
    # No duplicate tests and build the module name
    my (%seen, @test_files);
    foreach (@tests_to_run) {
        if (!$seen{$_}++) {
            # Does the test exist?
            if (!defined $self->available_tests->{$_}) {
                warn "$_ does not appear on $self->yaml_file";
            } else {
                if (!defined $self->available_tests->{$_}->{'code'}) {
                    warn "No code was defined for $_ on $self->yaml_file";
                } else {
                    my $description = uc $_;
                    if (defined $self->available_tests->{$_}->{'description'}) {
                        $description = $self->available_tests->{$_}->{'description'};
                    }
                    push(@test_files, [
                            $self->available_tests->{$_}->{'code'},
                            $description]);
                }
            }
        }
    }
    $self->prepare_environment();
    my $formatter_class = Test::System::Output::Factory->get_registered_class(
            $self->format
            );
    if (!defined $options->{'formatter_class'} && !defined $options->{'formatter'}) {
        $options->{'formatter_class'} = $formatter_class;
    }
    $options->{'merge'} = 0 unless defined $options->{'merge'};
    my $verbosity = -9;
    if (defined $options->{'verbosity'}) {
        $verbosity = $options->{'verbosity'};
        delete $options->{'verbosity'};
    }
    my $harness = TAP::Harness->new($options);
    $harness->formatter->verbosity($verbosity);
    $harness->runtests(@test_files);
    $self->clean_environment();
    return 1;
}

=item B<prepare_environment>

Prepares the environment by settings the needed environment values so they can
be used later by the tests

=cut
sub prepare_environment {
    my ($self) = @_;

    # Nodes are stored under the TEST_SYSTEM_NODES environment key
    $ENV{TEST_SYSTEM_NODES} = $self->nodes;
    if ($self->parameters) {
        if (ref \$self->parameters eq 'HASH') {
            foreach my $k (keys %{$self->parameters}) {
                # We are very cautious with environments variables already set
                $k = uc $k;
                if (!$ENV{'TEST_SYSTEM_' . $k}) {
                    my $value = $self->parameters->{$k};
                    if (ref \$value eq 'SCALAR') {
                        $ENV{'TEST_SYSTEM_' . $k} = $value;
                    } elsif (ref \$value eq 'ARRAY') {
                        $ENV{'TEST_SYSTEM_' . $k} = join(',', $value);
                    }
                }
            }
        }
    }
}

=item B<clean_environment>

Cleans/deletes all the environment variables that match I<TEST_SYSTEM_*>

=cut
sub clean_environment {
    my ($self) = @_;

    my %environment_vars = %ENV;
    foreach my $k (%environment_vars) {
        if ($k =~ /^TEST_SYSTEM_/) {
            delete $ENV{$k};
        }
    }
}

################################# Triggers ##################################
# Trigger for test_dir, when test_dir gets modified we look for all tests
# available in the given directory and so on we populate/fill the tests list
# attribute
sub _read_tests_yaml_file {
    my ($self, $yaml_file) = @_;
    
    if (!-f $yaml_file) {
        confess "The YAML file ($yaml_file) does not exist";
    }

    $self->{tests_yaml} = $yaml_file;
    $self->{tests_dir} = dirname($yaml_file);

    my $tests = LoadFile($yaml_file);
    foreach my $test (keys %$tests) {
        if (defined $tests->{$test}->{'code'}) {
            my $code = $tests->{$test}->{'code'};
            if (!-f $code) {
                $tests->{$test}->{'code'} = $self->{tests_dir} . '/' . $code;
            }
        }
    }
    $self->{available_tests} = $tests;
}

# Trigger for nodes, when nodes gets modified we don't really know if its
# a string or a list (and this is because we accept both) so we should make sure
# of what we get and validate it.
sub _verify_nodes_datatype {
    my ($self, $nodes) = @_;

    if (ref \$nodes eq 'ARRAY') {
        $self->{nodes} = join(',', $nodes);
    } else {
        $self->{nodes} = $nodes;
    }
}

# Trigger for format, when format gets modified we want to make sure the format
# is valid
sub _verify_format {
    my ($self, $format) = @_;

    if (!defined $self->available_formats->{$format}) {
        confess "The format you provided ($format) is not valid";
    }

    $self->{format} = $format;
}

# Generates the output type hash reference by checking the registered factory
# types of Test::System::Output::Factory
sub _generate_output_types_hash {
    my ($self) = @_;

    my @registered_types = Test::System::Output::Factory->get_registered_types;

    my %hash;
    foreach (@registered_types) {
        $hash{$_} = 1;
    }
    return \%hash;
}

1;



Test/System/Test.pm
#
# Test::System::Test
#
# Author(s): Pablo Fischer (pfischer@cpan.org)
# Created: 11/08/2009 14:03:24 PST 14:03:24
package Test::System::Test;

=head1 NAME

Test::System::Test - Desired parent class for all Test::System tests

=head1 DESCRIPTION

The purpose of this module is to provide the easiness of getting the list of
hostnames you want to test (if that is the case).

=cut

use Moose;
use Test::Class;
use Test::More;
use base qw(Test::Class);

=head2 Attributes

=over4

=item B<nodes>

A list of all the nodes/hosts that will be tested.

=over4
=cut
has 'nodes' => (
        is => 'ro',
        isa => => 'ArrayRef[Str]',
        default => \&generate_node_list
        );

# Fills the nodes attribute by checking the TEST_SYSTEM_NODES environment
# variable
sub generate_node_list {
    my $self = shift;
    my @node_list;
    if ($ENV{TEST_SYSTEM_NODES}) {
        @node_list = split(',', $ENV{TEST_SYSTEM_NODES});
    }
    # We don't like duplicated nodes..
    my %seen;
    my @unique = grep { ! $seen{$_}++ } @node_list;
    $self->{nodes} = \@unique;
}

######################### PRIVATE METHODS ############################
# Makes sure that before we start we have the needed information before
# starting the tests.
sub __setup_tests : Test(setup) {
    my $self = shift;

    if (!$self->nodes) {
        $self->SKIP_ALL("No nodes were given, nothing to test against");
    }
}

1;



example/suite.pl
#!/usr/bin/perl
# 
# suite.pl
# 
# Author(s): Pablo Fischer (pfischer@cpan.org)
# Created: 11/08/2009 15:31:37 PST 15:31:37

use strict;
use warnings;
use Getopt::Awesome qw(:all);
use Test::System;
use Test::System::Output::Factory;
use Data::Dumper;
require "example/MyFactory.pm";

define_option('test_groups=s', 'Test groups');
define_option('node=s@', 'Test nodes');
define_option('test_plan=s', 'Test plan');
define_option('test=s@', 'Test, can be a name or a file');
define_option('param=s%', 'Parameters we want to send');
define_option('no-warnings', 'Show warnings?');

my $suite = Test::System->new;
$suite->format_factory_class('MyFactory');

my @formats = keys(%{$suite->available_formats});
define_option('format=s', 'A valid format (' . join(', ', @formats) . ')');
parse_opts();

$suite->show_warnings(!get_opt('no-warnings'));

# Any nodes?
if (get_opt('node')) {
    $suite->nodes(get_opt('node'));
}

if (get_opt('test_groups')) {
    $suite->test_groups(get_opt('test_groups'));
}

my @tests;
if (get_opt('test')) {
    # An array ref!
    my $tmp_tests = get_opt('test');
    foreach (@$tmp_tests) {
        if (m/(.+?)#(.+?)$/) {
            push(@tests, [$1, $2]);
        } else {
            push(@tests, $_);
        }
    }
}

my $plan;
if (get_opt('test_plan')) {
    if (@tests) {
        warn "Hm, you provide a test_plan but also a test?!";
    } else {
        $plan = get_opt('test_plan');
    }
}

if (get_opt('format')) {
    $suite->format(get_opt('format'));
}

if (get_opt('param')) {
    $suite->parameters(get_opt('param'));
}

my %options;
$options{'merge'} = 'foo';
print Dumper($suite);
if ($plan) {
    $suite->run_test_plan($plan, \%options);
} else {
    $suite->runtests(\@tests, \%options);
}

# Results?
#print Dumper($suite);
